File: macos/PackMan/Sources/PackMan/Services/UpdateHistoryStore.swift
import Foundation

enum HistoryOutcome: String, Codable, CaseIterable, Sendable {
    case queued = "Queued", running = "Running", updated = "Updated", verified = "Verified"
    case failed = "Failed", unverified = "Unverified", cancelled = "Cancelled"
    case notStarted = "Not started", interrupted = "Interrupted", external = "Awaiting external update"
}

struct UpdateHistoryEntry: Identifiable, Codable, Sendable {
    var id = UUID()
    var runID: UUID
    var startedAt = Date.now
    var finishedAt: Date?
    var sourceID: SourceID
    var packageID: String
    var registryID: String?
    var name: String
    var beforeVersion: String
    var targetVersion: String
    var installedVersion: String?
    var outcome: HistoryOutcome = .queued
    var toolPath: String
    var evidence = ""
    var output = ""
    var verificationOnly = false

    var versionText: String { "\(beforeVersion) → \(installedVersion ?? "not confirmed") (requested \(targetVersion))" }
}

protocol UpdateHistoryStoring: Sendable {
    var loadIssue: String? { get }
    func read() -> [UpdateHistoryEntry]
    func save(_ entries: [UpdateHistoryEntry]) throws
}

final class UpdateHistoryStore: UpdateHistoryStoring, @unchecked Sendable {
    static let defaultURL = FileManager.default.homeDirectoryForCurrentUser
        .appendingPathComponent("Library/Application Support/PackMan/update-history.json")
    static let outputLimit = 64 * 1024
    private let url: URL?
    private let lock = NSLock()
    private var entries: [UpdateHistoryEntry] = []
    private(set) var loadIssue: String?

    init(url: URL? = defaultURL) {
        self.url = url
        guard let url, FileManager.default.fileExists(atPath: url.path) else { return }
        do {
            entries = try JSONDecoder().decode([UpdateHistoryEntry].self, from: Data(contentsOf: url))
            entries = Array(entries.prefix(500)).map { original in
                var entry = original
                if entry.outcome == .running || entry.outcome == .queued {
                    entry.outcome = entry.outcome == .running ? .interrupted : .notStarted
                    entry.evidence = "PackMan closed before this attempt completed. Verify before retrying."
                    entry.finishedAt = .now
                }
                entry.output = Self.bounded(entry.output)
                return entry
            }
        } catch {
            loadIssue = "History could not be read: \(error.localizedDescription). The original file has been retained."
        }
    }

    func read() -> [UpdateHistoryEntry] { lock.withLock { entries } }

    func save(_ updates: [UpdateHistoryEntry]) throws {
        try lock.withLock {
            var next = entries
            for var entry in updates {
                entry.output = Self.bounded(entry.output)
                if let index = next.firstIndex(where: { $0.id == entry.id }) { next[index] = entry }
                else { next.insert(entry, at: 0) }
            }
            next.sort { $0.startedAt > $1.startedAt }
            next = Array(next.prefix(500))
            if let url {
                let directory = url.deletingLastPathComponent()
                try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true,
                    attributes: [.posixPermissions: 0o700])
                // Preserve unreadable history before replacing it with a fresh store.
                if loadIssue != nil, FileManager.default.fileExists(atPath: url.path) {
                    let backup = url.deletingPathExtension().appendingPathExtension("unreadable-\(UUID().uuidString).json")
                    try FileManager.default.copyItem(at: url, to: backup)
                    try FileManager.default.setAttributes([.posixPermissions: 0o600], ofItemAtPath: backup.path)
                }
                let encoded = try JSONEncoder().encode(next)
                try encoded.write(to: url, options: .atomic)
                try FileManager.default.setAttributes([.posixPermissions: 0o600], ofItemAtPath: url.path)
            }
            entries = next
            loadIssue = nil
        }
    }

    static func bounded(_ text: String) -> String {
        guard text.utf8.count > outputLimit else { return text }
        return "[Earlier output truncated]\n" + String(decoding: text.utf8.suffix(outputLimit), as: UTF8.self)
    }
}

enum DiagnosticRedactor {
    static func redact(_ value: String, home: String = FileManager.default.homeDirectoryForCurrentUser.path) -> String {
        var text = value.replacingOccurrences(of: home, with: "~")
        // Scrub both assignments/headers and quoted JSON values, URL credentials,
        // common token forms, and private key blocks. Arbitrary output still merits review.
        let patterns: [(String, String)] = [
            (#"(?is)-----BEGIN [^-]*PRIVATE KEY-----.*?-----END [^-]*PRIVATE KEY-----"#, "[REDACTED PRIVATE KEY]"),
            (#"(?i)(https?://)[^\s/@]+:[^\s/@]+@"#, "$1[REDACTED]@"),
            (#"(?i)(authorization\s*[:=]\s*)(?:bearer|basic)\s+[^\s,;]+"#, "$1[REDACTED]"),
            (#"(?i)((?:[\"']?)(?:_authToken|_auth|password|passwd|secret|token|api[-_]?key|access[-_]?key|client[-_]?secret)(?:[\"']?)\s*[:=]\s*)(?:\"[^\"]*\"|'[^']*'|[^\s&;,]+)"#, "$1[REDACTED]"),
            (#"(?i)([?&](?:[^=&\s]*(?:token|secret|signature|credential|key|sig)[^=&\s]*)=)[^&\s]+"#, "$1[REDACTED]"),
            (#"\b(?:gh[pousr]_[A-Za-z0-9_]+|github_pat_[A-Za-z0-9_]+|npm_[A-Za-z0-9]+|sk-[A-Za-z0-9_-]{12,})\b"#, "[REDACTED]")
        ]
        for (pattern, replacement) in patterns {
            text = text.replacingOccurrences(of: pattern, with: replacement, options: .regularExpression)
        }
        return text
    }

    static func export(history: [UpdateHistoryEntry], log: [LogEntry]) throws -> String {
        // Redact fields before encoding so newlines and escaped JSON cannot bypass matching.
        let records = history.map { original in
            var entry = original
            entry.packageID = redact(entry.packageID); entry.registryID = entry.registryID.map { redact($0) }
            entry.name = redact(entry.name); entry.beforeVersion = redact(entry.beforeVersion)
            entry.targetVersion = redact(entry.targetVersion); entry.installedVersion = entry.installedVersion.map { redact($0) }
            entry.toolPath = redact(entry.toolPath); entry.evidence = redact(entry.evidence); entry.output = redact(entry.output)
            return entry
        }
        struct Export: Encodable { let version: String; let history: [UpdateHistoryEntry]; let log: [String] }
        let encoder = JSONEncoder(); encoder.outputFormatting = [.prettyPrinted, .sortedKeys]; encoder.dateEncodingStrategy = .iso8601
        let data = try encoder.encode(Export(version: AppViewModel.appVersion, history: records,
            log: log.map { redact("\($0.timestamp) [\($0.scope ?? "PackMan")] \($0.message)") }))
        return String(decoding: data, as: UTF8.self)
    }
}

File: windows/src/PackMan/Services/UpdateHistoryStore.cs
using System.Text.Json;
using System.Text.RegularExpressions;

namespace PackMan.Services;

public enum HistoryOutcome { Running, Updated, Failed, Cancelled, Unverified, NotStarted, Interrupted, Verified }

public sealed record UpdateHistoryEntry(
    Guid Id, Guid RunId, DateTimeOffset StartedAt, DateTimeOffset? FinishedAt,
    SourceId Source, string PackageId, string Name, string? Repository, string BeforeVersion,
    string TargetVersion, string? InstalledVersion, HistoryOutcome Outcome, RestartState Restart,
    string? Evidence, string? Message, string ToolPath, string Output)
{
    public string OutcomeText => Outcome + (Restart == RestartState.None ? "" : $" • restart {Restart.ToString().ToLowerInvariant()}");
    public string VersionText => $"{BeforeVersion} → {InstalledVersion ?? "not confirmed"} (requested {TargetVersion})";
    public string Details => $"{Name} ({PackageId})\n{Source} / {Repository ?? "default repository"}\n{VersionText}\n"
        + $"{OutcomeText}\n{StartedAt:g} — {FinishedAt:g}\n{ToolPath}\n{Evidence}\n{Message}\n\n{Output}";
}

public interface IUpdateHistoryStore
{
    string? LoadIssue { get; }
    IReadOnlyList<UpdateHistoryEntry> Read();
    void Save(UpdateHistoryEntry entry);
}

public sealed class UpdateHistoryStore : IUpdateHistoryStore
{
    private readonly string? _path;
    private readonly object _sync = new();
    private List<UpdateHistoryEntry> _entries = [];
    public string? LoadIssue { get; private set; }
    public static string DefaultPath => Path.Combine(Environment.GetFolderPath(
        Environment.SpecialFolder.LocalApplicationData), "PackMan", "update-history.json");

    public UpdateHistoryStore(string? path)
    {
        _path = path;
        if (path is null || !File.Exists(path)) return;
        try
        {
            _entries = (JsonSerializer.Deserialize<List<UpdateHistoryEntry>>(File.ReadAllText(path)) ?? [])
                .Take(500).Select(entry => entry.Outcome == HistoryOutcome.Running
                    ? entry with { Outcome = HistoryOutcome.Interrupted,
                        Message = "PackMan closed before the result was recorded. Scan or verify before retrying." }
                    : entry).ToList();
        }
        catch (Exception ex) { LoadIssue = $"Update history could not be loaded: {ex.Message}"; }
    }

    public IReadOnlyList<UpdateHistoryEntry> Read() { lock (_sync) return _entries.ToArray(); }

    public void Save(UpdateHistoryEntry entry)
    {
        entry = entry with { Output = DiagnosticRedactor.Redact(entry.Output),
            Message = DiagnosticRedactor.Redact(entry.Message), Evidence = DiagnosticRedactor.Redact(entry.Evidence) };
        lock (_sync)
        {
            var next = _entries.Where(e => e.Id != entry.Id).Prepend(entry).Take(500).ToList();
            if (_path is not null)
            {
                Directory.CreateDirectory(Path.GetDirectoryName(_path)!);
                var temporary = _path + "." + Guid.NewGuid().ToString("N") + ".tmp";
                try
                {
                    File.WriteAllText(temporary, JsonSerializer.Serialize(next, new JsonSerializerOptions { WriteIndented = true }));
                    File.Move(temporary, _path, true);
                }
                finally { if (File.Exists(temporary)) File.Delete(temporary); }
            }
            _entries = next;
        }
    }
}

public static class DiagnosticRedactor
{
    public static string Redact(string? value)
    {
        if (string.IsNullOrEmpty(value)) return string.Empty;
        var text = Regex.Replace(value, @"(?i)(https?://)[^\s/@]+:[^\s/@]+@", "$1[redacted]@");
        text = Regex.Replace(text, @"(?i)(Bearer\s+)\S+", "$1[redacted]");
        text = Regex.Replace(text, @"(?i)((?:[_a-z-]*(?:token|password|passwd|secret|apikey|api_key|api-key|authorization)[_a-z-]*)[\""']?\s*[:=]\s*[\""']?)[^\s,;\""'&]+", "$1[redacted]");
        text = Regex.Replace(text, @"([?&][^\s=&]+)=([^\s&#]+)", "$1=[redacted]");
        var profile = Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
        if (!string.IsNullOrWhiteSpace(profile)) text = text.Replace(profile, "%USERPROFILE%", StringComparison.OrdinalIgnoreCase);
        return text;
    }
}
